### PR TITLE
Feature/auth

### PR DESCRIPTION
**Pull request by retrokiller543**

### CODE REVIEW

File: .env.local
Comment (writemorecode) [inline .env.local:1]: Using the "root" user will work for now, but we will need to change this to a user with fewer privs later.

File: src/auth/mod.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+pub(crate) enum Role {
+    Owner,
+    Admin,
+    #[default]
+    User,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialOrd, Eq, PartialEq, Clone)]
+pub(crate) struct UserInfo {
+    pub(crate) id: Option<Thing>,
```
Comment (writemorecode) [inline src/auth/mod.rs:21]: Why is the user ID an Option type? Is it possible for a user to not have an ID?

File: src/auth/mod.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+use crate::{PaginationResponse, Record, INTERNAL_DB};
+use anyhow::bail;
+use serde::{Deserialize, Serialize};
+use surrealdb::sql::{Datetime, Thing};
+
+pub mod oauth;
+pub mod session;
+pub mod users;
+
+#[derive(Debug, Serialize, Deserialize, PartialOrd, Eq, PartialEq, Clone, Default)]
```
Comment (writemorecode) [inline src/auth/mod.rs:11]: What is the difference between an Owner and an Admin? Can there be multiple owners?

File: src/auth/mod.rs (hunk 1 above)
Comment (retrokiller543) [reply src/auth/mod.rs:21]: This is so we can add a user and let the db create a optimal Id, without option we would need to make a id ourselves

File: .env.local
Comment (retrokiller543) [reply .env.local:1]: Agreed, this is linked to what we do in the dev repo

File: src/auth/mod.rs (hunk 2 above)
Comment (retrokiller543) [reply src/auth/mod.rs:11]: Technically yes there can and there is a difference between a admin and a owner but we can adjust naming and permission levels as it becomes more relevant

File: src/auth/mod.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    #[serde(flatten)]
+    pub(crate) pagination: PaginationResponse,
+}
+
+#[tracing::instrument(skip(password))]
+async fn create_auth_for_user(
+    user_id: Record,
+    providers: Vec<OauthProvider>,
+    password: Option<String>,
```
Comment (writemorecode) [inline src/auth/mod.rs:45]: Is this a plaintext or a hashed password? Naming could be more clear.

Comment (retrokiller543) [reply src/auth/mod.rs:45]: This is not implemented yet and only placeholder, the naming I think is accurate as it's a password, then if it's hashed or not is of little importance for the name. And a string is probably the best type unless we want a custom solution

File: src/auth/users/get/mod.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    if let Some(user) = user {
+        Ok(user)
+    } else {
+        Err(anyhow::anyhow!("User not found"))
+    }
+}
+
+#[tracing::instrument(skip(db, email))]
+pub(crate) async fn get_user<T>(db: &Arc<Surreal<T>>, email: &str) -> Option<UserInfo>
```
Comment (writemorecode) [inline src/auth/users/get/mod.rs:53]: Do we need this function? Is it just the "get_user_by_email" function with another name?

File: src/auth/users/get/mod.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub username: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub email: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub token: Option<String>,
+}
+
+#[tracing::instrument(skip(db, data))]
+pub(crate) async fn get_user_by_internal<T>(
```
Comment (writemorecode) [inline src/auth/users/get/mod.rs:71]: Great idea!

File: src/auth/users/get/mod.rs (hunk 4 above)
Comment (retrokiller543) [reply src/auth/users/get/mod.rs:53]: Unsure, think i had it in my old code and never used it but thought of using it

File: src/auth/mod.rs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+) -> anyhow::Result<()> {
+    let mut require_password = false;
+    let providers_ids = providers
+        .iter()
+        .map(|p| p.id.clone())
+        .collect::<Vec<Thing>>();
+    let user_id = user_id.id;
+
```
Comment (writemorecode) [inline src/auth/mod.rs:53]: Maybe we could do something like:
`let require_password = providers.iter().any(|p| p.name == OauthProviderName::Email);`.

Comment (retrokiller543) [reply src/auth/mod.rs:53]: Yes that could be done bit we havet to remeber that the code for the Email provider does not yet exist and will most likely change when we get to it so optimizing and cleaning might not be needed at this stage. This PR is mainly so we can get this code into the dev repo and setup docker compose, even if the auth glow is very much in early stages and has code that can be improved.

**Review by kebabohvo — APPROVED** [review]

**Pull request merged**